spopt/region/skater.py

```diff
@@ -0,0 +1,409 @@
+            local_MSF[in_node, out_node] = 0
+            local_MSF.eliminate_zeros()
+
+            # get the connected components
+            local_n_subtrees, local_labels = cg.connected_components(
+                local_MSF, directed=False
+            )
+
+            if local_n_subtrees <= current_n_subtrees:
+                raise Exception("Malformed MSF!")
+
+            # compute the score of these components
+            score = self.score(data, labels=local_labels, quorum=quorum)
+
+            # if the score is lower than the best score and quorum is met
+            if score < best_deletion.score:
+                best_deletion = deletion(in_node, out_node, score)
+        if make:
+            return self.make_cut(*best_deletion, MSF=MSF)
+        return best_deletion
+
+    def make_cut(self, in_node, out_node, score, MSF=None):
+        """
+        make a cut on the MSF inplace, provided the in_node, out_node, MSF, and score. 
+        in_node: int, ID of the source node for the edge to be cut
+        out_node: int, ID of the destination node for the edge to be cut
+        score: float, the value of the score being cut. if the score is infinite, the cut is not made. 
+        MSF: the spanning forest to use when making the cut. If not provided,
+             uses the defualt tree in self.minimum_spanning_forest_
+        """
+        if MSF is None:
+            MSF = self.minimum_spanning_forest_
+        if np.isfinite(score):
+            MSF[in_node, out_node] = 0
+            MSF.eliminate_zeros()
+            return (MSF, *cg.connected_components(MSF, directed=False))
+        raise OptimizeWarning(
+            "Score of the ({},{}) cut is inf, the quorum is likely not met!"
+        )
+
+
+class Skater(BaseSpOptHeuristicSolver):
+    """...Needs a short description..."""
+
+    def __init__(
+        self,
+        gdf,
+        w,
+        attrs_name,
+        n_clusters=5,
+        floor=-np.inf,
+        trace=False,
+        islands="increase",
+    ):
+        """
+        # SKATER spatial clustering algorithm.
+        Parameters
+        ----------
+        
+        gdf : geopandas.GeoDataFrame
+            ... : 
+   
+        w : libpysal.weights.W instance
+        spatial weights matrix
+
+        attrs_name : list
+        Strings for attribute names (cols of ``geopandas.GeoDataFrame``).
+        
+        n_clusters : int, optional, default: 5
+        The number of clusters to form.
+
+        floor : type
+        TODO.
+
+        trace : type
+        TODO.
+
+        islands : type
+        TODO.
+        
+        """
+        self.gdf = gdf
+        self.w = w
+        self.attrs_name = attrs_name
+        self.n_clusters = n_clusters
+        self.floor = floor
+        self.trace = trace
+        self.islands = islands
+
+    def solve(self):
+        """Solve the region k-means heuristic."""
+        data = self.gdf
+        X = data[self.attrs_name].values
+        model = SpanningForest()
```

Comment:
Shouldn't users be able to pass in the keyword arguments to `SpanningForest.__init__()` here (e.g. `dissimilarity`)?